### DISCUSSION TITLE
Object duplication ideas?

### DISCUSSION THREAD

**Discussion by kaveh808** [kaveh808/kons-9#62] [Ideas]
Anyone have any ideas on how to handle object duplication for scene items in a general way?

The manual way would be to use a combination of `duplicate-object (src)` and `copy-object-data (dst src)`.

**Comment by mikelevins** [comment]
Depends on what you mean, but I can think of two different approaches, one drawn from prototype object systems like SK8's, and one from Entity-Component-System architectures as used in many games.

The SK8-like approach (derived from CycL, which is derived from Minsky frames, and which influenced systems like Self and through that Newtonscript and Javascript) is that you have a database of prototype objects that you "instantiate" by copying and then updating slot values.

You can reduce the overall copying at the cost of a little additional architectural complexity by having prototype relationships of different kinds (SK8 did this, as did CycL): you copy a template object and then assign, for example, a structure prototype from which it inherits its arrangement of slots along with default values, and a behavior prototype, from which it inherits its behavior (that is, methods), and a containment prototype, from which it inherits certain behavior, such as event handling, based on what contains it, and a truth-maintenance prototype, from which it inherits invariants that it must maintain, and so on.

Designing this system is hairy, but the result is objects that are simple and flexible and that inherit a lot of useful stuff automatically.

The entity-component-system gets a similar effect a different way: you have entities, which are template objects that you clone to make new instances. Then you have components, which are small, simple objects that add specific traits and behaviors one at a time. You add them to entities as-needed. Then you have systems, which are coordinator objects. You add entities to them and they organize cross-object behaviors.

In each case the idea is to share as much structure as you can, while supporting the ability to reuse large numbers of compact copies of (or proxies for) template objects.

**Comment by kaveh808** [comment]
On one level there is the issue of duplicating a self-contained item, such as a polyhedron. The points array, face array and other slots all need to be duplicated as using a deep copy mechanism.

In a more complex example, the user may want to duplicate a group in the scene graph which may have children and dependencies to other objects as well as possibly function pointers for internal actions.

It is possible (but tedious and error prone) to manually write duplication methods for every class, and then recursively call them as necessary. I'm wondering if there is some sort of automated way of doing some of this work.

Worst case we write it all by hand, which is not the end of the world. Just have to be careful and methodical about it.

**Reply by mikelevins** [reply]
Right; the point of ECS or the parenting system is to give you a framework for automating that kind of thing. WIth ECS you add some "deep-copy" trait  or something that provides the copying behavior you want. With the frame-based approach, you give them a "copy parent" or something similar. In either case, you only have to write the copying code once per kind of copying algorithm (assuming you include a means for marking which slots you want to copy).

The nice thing about either of these approaches, as opposed to writing copying methods ad hoc by hand, is that if you can look at the copy parent, or the copy trait, that object tells you what kind of copying gets done. If you do it ad hoc per object, then you have to look at each and every object's copy code to figure out what it does.

**Comment by kaveh808** [comment]
I need a better understanding of these approaches. Maybe what I will do is implement the straightforward way for the kernel classes and then we can look at them and re-factor to one or other of the approaches before we extend duplication to the plugin classes.

**Reply by mikelevins** [reply]
ECS is simpler to implement and easier to understand quickly. The basic idea is that your already-existing data objects are the entities, systems are control programs that you want to apply to them, and components are properties or behaviors that you can add to your entities to control what the systems do.

So we can imagine that you have a simple copying system. Its job is to make copies of input objects. You have a few different ways to copy things, but you don't want to descend to writing an entire separate copying algorithm for each and every different type of object.

So, just for simplicity's sake, let's suppose your copying system consists of deep-copy and shallow-copy, and you have a generic driver function called copy. To make some data structure participate in the copying system, you add a trait to it. To control which algorithm gets applied, you add either DEEP-COPYABLE or SHALLOW-COPYABLE.

Maybe you have cases where you want some but not all slots to get copied. Then you might want to add a trait SLOTS-TO-COPY that can be :ALL or a list of slot names.

I'm handwaving one important detail: how do you add traits to entities? It depends on what the data objects are. If they're CLOS objects, you can make them inherit from an ENTITY mixin that provides slots for managing traits. If they're lighter weight and you want to keep them that way, then you might implement entities as wrapper objects: you instantiate an array entity by making an ARRAY-ENTITY instance that contains the actual array and provides the hooks that the ECS architecture needs to manage traits and apply system functions.

If you need things to be even lighter-weight than that, then you can implement entities as collections or containers; in that case you make your polyhedron into an entity by adding it to a POLYHEDRON-ENTITIES collection, and the collection object takes care of the entity aspect of the objects it contains. (If you use that last strategy, then obviously all the values in the container must have the same entity-related properties, since the properties live on the container.)

**Comment by JMC-design** [comment]
It seems this is a question more easily answered with core profile and things like gl:draw-arrays/elements-instanced.
Then you only need to store deltas

**Comment by loskool** [comment]
A "hacky" way to do it, though slow, would be to use any of the already-existing object serialization tools to serialize and deserialize your objects in one go. You'd end up with a different set of objects, occupying different memory, that all have the same "internal references" to one another as did the original set of objects. That is if Object A referred to object B, then object A' would refer to object B'.

Or at the very least, something like cl-store used as a guide to implement a generic deep copying mechanism.

**Reply by kaveh808** [reply]
Interesting idea.

Do we then need to make a distinction between an object B that is owned by object A, versus an object B that is reference by object A and should not be duplicated?

**Reply by mikelevins** [reply]
Probably. It depends on how the application uses the object graphs, but every object persistence system I’ve used includes a way to say which slots get copied and which don’t, because it’s just such a common thing to need to keep track of.

**Comment by aykaramba** [comment]
Am I correct to understand that the ECS approach would provide the project with almost for free non destructive editing and history?

**Reply by mikelevins** [reply]
No. 

You may be confusing ECS, which is a way to organize shared properties and behaviors, with object prevalence, which is a way to implement object persistence.  We've discussed both ECS and prevalence.

Prevalence works by maintaining an in-memory data structure of objects to be saved, and making changes to them only by using APIs that log them to a journal. That logging process also provides a convenient hook for implementing history and undo.

**Reply by aykaramba** [reply]
Thank you.

**Comment by kaveh808** [comment]
I pushed an initial implementation of object duplication in `scene-duplicate.lisp`

Test case "Demo 16" in `demo-kernel.lisp`

Needs more testing.
